### PR TITLE
Dylan handle gzip encoding properly

### PR DESCRIPTION
**Pull request by dylan-tock**
Adding detection of `Accept-Header: .*gzip.*` in the client request.
* If the header is present and contains `gzip`, send on the compressed version of the file retrieved from GCS and use the compressed size as the `Content-Length` reply header.
* If the header is not present or `gzip` is not present within the header, get the uncompressed version of the file from GCS and do not set a `Content-Length` reply header.

Also made these changes:
* Update to go 1.18
* Update imports (via `go mod tidy`)
* Update `Dockerfile` to use the "use builder image to create binary; copy binary to service image for deployment" pattern.
* Update `.gitignore` to ignore any `gcsproxy` binaries that are built during testing; and to ignore vim swap files.

### CODE REVIEW

**Comment by dylan-tock** [comment]
Tested using the following process:

```
curl -sSL -v http://localhost:8080/tock-jo2y-assets/instrument.js | tee js1
echo ""
echo "Above without gzip"
sleep 1
curl -sSL -v -H "accept-encoding: gzip, deflate, br" http://localhost:8080/tock-jo2y-assets/instrument.js | gzip -dc - | tee js2
echo ""
echo "Above WITH gzip"
diff js1 js2 && echo "No Differences detected"
```

File: main.go [hunk 1]
```diff
@@ -128,10 +130,20 @@ func proxy(w http.ResponseWriter, r *http.Request) {
...
-	setIntHeader(w, "Content-Length", attr.Size)
+	if gzipAcceptable {
+		setIntHeader(w, "Content-Length", attr.Size)
+	}
 	io.Copy(w, objr)
 }
 
+func clientAcceptsGzip(r *http.Request) bool {
+	acceptHeader := r.Header.Get("Accept-Encoding")
+	if strings.Contains(acceptHeader, "gzip") {
```
Comment (jjlueck-tock) [inline main.go]: I believe this can just be `return strings.Contains(...)`

**Review by jjlueck-tock — COMMENTED** [review]
Just one very minor thing from me...

File: main.go (hunk 1 above)
Comment (dylan-tock) [reply main.go]: You are absolutely correct. Tested it and works just peachy!

**Review by jjlueck-tock — APPROVED** [review]
Cool by me

**Review by jo2y-tock — APPROVED** [review]
Looks reasonable to me.
Some notes:
- I'm likely not going to use the Dockerfile for our use-case. I need the binary in the haproxy image. I'll take care of that in my next round of testing.
- If we plan to push this change upstream, I worry that some things might be too far reaching. Eg. I think the ubuntu image is larger than alpine.
- Not setting Content-Length for the uncompressed use-case is probably fine. But I think the value we do want in that use case is there, but on a different object. objr instead of obj maybe? It's been a while since I looked at the api.

**Comment by dylan-tock** [comment]
> Looks reasonable to me. Some notes:
> 
> * I'm likely not going to use the Dockerfile for our use-case. I need the binary in the haproxy image. I'll take care of that in my next round of testing.
> * If we plan to push this change upstream, I worry that some things might be too far reaching. Eg. I think the ubuntu image is larger than alpine.

I was thinking about that after I submitted and would like to actually remove those from the PR to make it more acceptable upstream.... I might do another PR with just the gzip fix and, if necessary/desired, put the other changes into a different PR.

> * Not setting Content-Length for the uncompressed use-case is probably fine. But I think the value we do want in that use case is there, but on a different object. objr instead of obj maybe? It's been a while since I looked at the api.

I'll take a look at this when I create the smaller PR.

**Pull request closed without merging**